hledger-lib/Hledger/Reports/BudgetReport.hs

```diff
@@ -47,8 +47,8 @@ type BudgetAverage = Average
 
 -- | A budget report tracks expected and actual changes per account and subperiod.
 type BudgetCell = (Maybe Change, Maybe BudgetGoal)
-type BudgetReport = PeriodicReport BudgetCell
-type BudgetReportRow = PeriodicReportRow BudgetCell
+type BudgetReport = PeriodicReport AccountLeaf BudgetCell
+type BudgetReportRow = PeriodicReportRow AccountLeaf BudgetCell
```

Comment:
I guess the benefit of this commit is that we can now typefully describe unlabelled rows ( with () ), and we have the possibility of attaching other kinds of labels.

I don't love the name AccountLeaf but I'm not thinking of anything better either.

Reply:
Typefully describing unlabelled rows is the application in the library itself. My motivation for adding this is that have a need for using other kinds of labels in my own program (an account name along with a category). I actually wrote my own version of a generalised MultiBalanceReport with other types of labels, complete with boilerplate and all, before realising the existence of PeriodicReport.

I also don't love the name AccountLeaf. I've moved back and forth between that and AccountTag, but I think AccountLeaf is a bit more descriptive of the standard use, and less verbose than AccountWithLeaf.

I actually wonder if the correct solution is to just store the full account name, and to calculate the leaf name when needed for display or other purposes.

Reply:
I'd be interested to see that. I'm ok with leaving it as-is for now if you prefer. I think we were discussing the missing feature of MultiBalanceReport compared to BalanceReport (eliding parent accounts), so perhaps this will affect it (perhaps you're thinking of that already).

Reply:
I also wonder whether the account depth belongs more properly in the AccountLeaf structure, or whether it should continue to live in PeriodicReport itself.

Reply:
I've put a bit of thought into how to add that missing feature. The main problem, as I see it, is that determining which accounts are boring involves at all columns, rather than just a single one. This leads to a messy problem, with a couple possible solutions, none of which seem ideal to me.

1. Check that each column has value zero, and that union of the subaccounts over all columns has a single element. (Do we consider subaccounts to be the same if they have the same account name? Will we run into any problems with that?) If so, it's boring.

2. Collapse the account names by going up the parents checking for boring accounts. This is actually rather troublesome, as:

    1. Different columns could conceivably have different parent accounts. Does this ever happen in practice?

    2. The parent of an account only corresponds to a single column, so in order to see whether it's boring, we need a separate lookup table, rather that just using `aboring`, like BalanceReport is able to do. We again run into the problem of whether using just the account name as the key for the lookup table is sufficient.

I've made a couple of steps in the direction of some solution or another, but they all seem to be a bit hacky to me, trying to work around limitations of the Account data type, which is designed with a single balance in mind.